lib/PipelineC/PipelineCTracing.cpp

```diff
@@ -0,0 +1,283 @@
+  }
+
+  std::string reprString(const char *String) {
+    return '"' + llvm::StringRef(String).str() + '"';
+  }
+
+public:
+  void functionPrelude(const llvm::StringRef Name) {
+    OS << "- name: " << Name << "\n";
+    OS << "  arguments:\n";
+    OutputtingArguments = true;
+    OS.flush();
+  }
+
+  void newArgument() {
+    OS << "  - ";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_same_v<T, uint8_t> ||
+           std::is_same_v<T, uint32_t> ||
+           std::is_same_v<T, uint64_t>
+  // clang-format on
+  void printValue(const T &Int) {
+    OS << Int << "\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_same_v<T, char *> ||
+           std::is_same_v<T, const char *>
+  // clang-format on
+  void printValue(T String) {
+    if (OutputtingArguments)
+      OS << reprString(String);
+    else
+      OS << "P" << (void *) String;
+    OS << "\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  requires std::is_same_v<T, bool>
+  void printValue(const bool &Bool) {
+    OS << (Bool ? "true" : "false") << "\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires (std::is_pointer_v<T> && isRPType<remove_constptr<T>>())
+  void printValue(T Ptr) {
+    // clang-format on
+    OS << "P" << (void *) Ptr << "\n";
+    OS.flush();
+  }
+
+  void printPointer(void *Ptr) {
+    // clang-format on
+    OS << "P" << Ptr << "\n";
+    OS.flush();
+  }
+
+  void printBuffer(const char *Buffer, const uint64_t Length) {
+    OS << llvm::encodeBase64(llvm::StringRef(Buffer, Length)) << "\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_same_v<T, uint8_t*> ||
+           std::is_same_v<T, uint32_t*> ||
+           std::is_same_v<T, uint64_t*>
+  // clang-format on
+  // Integer list printing
+  void printList(T IntList, const uint64_t &Length) {
+    OS << "[";
+    for (uint64_t I = 0; I < Length; I++) {
+      OS << (uint64_t) IntList[I];
+      if (I < Length - 1) {
+        OS << ", ";
+      }
+    }
+    OS << "]\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_same_v<T, char **> ||
+           std::is_same_v<T, const char **>
+  // clang-format on
+  void printList(T StringList, const uint64_t Length) {
+    OS << "[";
+    for (uint64_t I = 0; I < Length; I++) {
+      OS << reprString(StringList[I]);
+      if (I < Length - 1) {
+        OS << ", ";
+      }
+    }
+    OS << "]\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires (isRPType<remove_constptr<remove_constptr<T>>>())
+  void printList(T PtrList, const uint64_t Length) {
+    // clang-format on
+    OS << "[";
+    for (uint64_t I = 0; I < Length; I++) {
+      OS << "P" << (void *) PtrList[I];
+      if (I < Length - 1) {
+        OS << ", ";
+      }
+    }
+    OS << "]\n";
+    OS.flush();
+  }
+
+  void printVoid() {
```

Comment:
Use case?

Reply:
Functions that "return" void
It's needed so in the trace we get `return: null\n`